File: benchmarks/compare/raptor/go/filter.go
package raptor

import "sort"

// JourneyFilter filters a list of journeys.
type JourneyFilter interface {
	Apply(journeys []Journey) []Journey
}

// FilterCriteria returns true if b is at least as good as a in some dimension.
type FilterCriteria func(a, b Journey) bool

// EarliestArrival returns true if b arrives at or before a.
func EarliestArrival(a, b Journey) bool { return b.ArrivalTime <= a.ArrivalTime }

// LeastChanges returns true if b has the same or fewer changes than a.
func LeastChanges(a, b Journey) bool { return len(b.Legs) <= len(a.Legs) }

// MultipleCriteriaFilter sorts journeys and removes dominated ones.
type MultipleCriteriaFilter struct {
	criteria []FilterCriteria
}

// NewMultipleCriteriaFilter constructs a filter with the default criteria
// (earliestArrival + leastChanges).
func NewMultipleCriteriaFilter(criteria ...FilterCriteria) *MultipleCriteriaFilter {
	if len(criteria) == 0 {
		criteria = []FilterCriteria{EarliestArrival, LeastChanges}
	}
	return &MultipleCriteriaFilter{criteria: criteria}
}

// Apply sorts the journeys (departure asc, arrival desc tiebreak) and keeps
// each journey unless some later journey dominates it on all criteria.
func (f *MultipleCriteriaFilter) Apply(journeys []Journey) []Journey {
	sort.SliceStable(journeys, func(i, j int) bool {
		a, b := journeys[i], journeys[j]
		if a.DepartureTime != b.DepartureTime {
			return a.DepartureTime < b.DepartureTime
		}
		// arrival time descending tiebreak
		return a.ArrivalTime > b.ArrivalTime
	})

	out := []Journey{}
	for i := range journeys {
		if f.keep(journeys, i) {
			out = append(out, journeys[i])
		}
	}
	return out
}

func (f *MultipleCriteriaFilter) keep(journeys []Journey, index int) bool {
	a := journeys[index]
	for j := index + 1; j < len(journeys); j++ {
		b := journeys[j]
		dominated := true
		for _, c := range f.criteria {
			if !c(a, b) {
				dominated = false
				break
			}
		}
		if dominated {
			return false
		}
	}
	return true
}

File: benchmarks/compare/raptor/rust/src/filter.rs
//! Journey filters (port of src/results/filter/{JourneyFilter,MultipleCriteriaFilter}.ts).

use crate::journey::Journey;

/// Filter a number of journeys.
pub trait JourneyFilter {
    fn apply(&self, journeys: Vec<Journey>) -> Vec<Journey>;
}

/// b arrives before or at the same time as a.
fn earliest_arrival(a: &Journey, b: &Journey) -> bool {
    b.arrival_time <= a.arrival_time
}

/// b has the same or fewer changes than a.
fn least_changes(a: &Journey, b: &Journey) -> bool {
    b.legs.len() <= a.legs.len()
}

type FilterCriteria = fn(&Journey, &Journey) -> bool;

pub struct MultipleCriteriaFilter {
    criteria: Vec<FilterCriteria>,
}

impl Default for MultipleCriteriaFilter {
    fn default() -> Self {
        MultipleCriteriaFilter {
            criteria: vec![earliest_arrival, least_changes],
        }
    }
}

impl MultipleCriteriaFilter {
    pub fn new() -> Self {
        Self::default()
    }
}

impl JourneyFilter for MultipleCriteriaFilter {
    fn apply(&self, mut journeys: Vec<Journey>) -> Vec<Journey> {
        // Sort by departure asc; tie-break by arrival descending. Stable sort.
        journeys.sort_by(|a, b| {
            if a.departure_time != b.departure_time {
                a.departure_time.cmp(&b.departure_time)
            } else {
                b.arrival_time.cmp(&a.arrival_time)
            }
        });

        // Keep journey A unless some LATER journey B satisfies ALL criteria.
        let mut out = Vec::new();
        for i in 0..journeys.len() {
            let a = &journeys[i];
            let dominated = journeys[i + 1..]
                .iter()
                .any(|b| self.criteria.iter().all(|c| c(a, b)));
            if !dominated {
                out.push(journeys[i].clone());
            }
        }

        out
    }
}
